dhcpv4/option_archtype.go

```diff
@@ -53,24 +54,11 @@ func (o *OptClientArchType) String() string {
 // ParseOptClientArchType returns a new OptClientArchType from a byte stream,
 // or error if any.
 func ParseOptClientArchType(data []byte) (*OptClientArchType, error) {
-	if len(data) < 2 {
-		return nil, ErrShortByteStream
-	}
-	code := OptionCode(data[0])
-	if code != OptionClientSystemArchitectureType {
-		return nil, fmt.Errorf("expected code %v, got %v", OptionClientSystemArchitectureType, code)
-	}
-	length := int(data[1])
-	if length == 0 || length%2 != 0 {
-		return nil, fmt.Errorf("Invalid length: expected multiple of 2 larger than 2, got %v", length)
-	}
-	if len(data) < 2+length {
-		return nil, ErrShortByteStream
-	}
-	archTypes := make([]iana.ArchType, 0, length%2)
-	for idx := 0; idx < length; idx += 2 {
-		b := data[2+idx : 2+idx+2]
-		archTypes = append(archTypes, iana.ArchType(binary.BigEndian.Uint16(b)))
+	buf := uio.NewBigEndianBuffer(data)
+
+	archTypes := make([]iana.ArchType, 0, buf.Len()/2)
+	for buf.Has(2) {
+		archTypes = append(archTypes, iana.ArchType(buf.Read16()))
 	}
-	return &OptClientArchType{ArchTypes: archTypes}, nil
+	return &OptClientArchType{ArchTypes: archTypes}, buf.FinError()
```

Comment:
the RFC 4578 section 2.1 requires that at least one arch type is present. Here (or above) we should return an error if `len(archTypes)` is 0